### PR TITLE
Dockerize openttd

### PR DESCRIPTION
**Pull request by santhoshjanan**
## Commits
- [Dockerfile, example configuration file and documentation updated](https://github.com/OpenTTD/OpenTTD/commit/81b6aa603349db687e06ecd588ae2c1283d43589)
- [Updated readme with command to docker pull moriartee/openttd-server](https://github.com/OpenTTD/OpenTTD/commit/3ffe75b097a5fd8daa37d5a02b890b6ba50618dc)

## Motivation / Problem

I was trying to set up a dedicated server on an ubuntu server, without installing all the dev packages. So, I guess it will be a generic usecase to set up a completely customizable docker container.

## Description

* The dockerfile can be used to build any latest version of OpenTTD with minimum to no effort.
* All that we need to have is a copy of the openttd.cfg-example

## Limitations

* If the build file/path/patterns change, we may need to update the Dockerfile.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* This PR does not affect the gameplay or any in-game features.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+# ------------------------------------------------------
+ENV ZLIB_VERSION=1.3.1
+WORKDIR /tmp
+RUN curl -LO "https://www.zlib.net/zlib-${ZLIB_VERSION}.tar.gz" \
+    && tar zxvf "zlib-${ZLIB_VERSION}.tar.gz" \
+    && cd "zlib-${ZLIB_VERSION}" \
+    && ./configure --prefix=/usr/local \
+    && make -j"$(nproc)" \
+    && make install \
+    && cd / && rm -rf /tmp/zlib*
```
Comment (glx22) [inline Dockerfile:60]: Why not just install `zlib1g-dev` with `apt`?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+# --------------------------------------------------------
+WORKDIR /build
+RUN git clone --depth=1 --branch ${APPVER} https://github.com/OpenTTD/OpenTTD.git
+
+# --------------------------------------------------------
+# 5) Build OpenTTD (Release Mode)
+# --------------------------------------------------------
+WORKDIR /build/OpenTTD
+RUN mkdir build && cd build && \
+    cmake .. -DCMAKE_BUILD_TYPE=Release &&\
```
Comment (glx22) [inline Dockerfile:90]: Since you are running as dedicated server, don't do a full build.
```suggestion
    cmake .. -DCMAKE_BUILD_TYPE=Release -DOPTION_DEDICATED=ON &&\
```

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    # OpenTTD optional/encouraged libraries (dev packages, except zlib):
+    liblzma-dev \
+    libpng-dev \
+    liblzo2-dev \
+    libfreetype6-dev \
+    libfontconfig1-dev \
+    libicu-dev \
+    libharfbuzz-dev \
+    libcurl4-openssl-dev \
+    libsdl2-dev \
```
Comment (glx22) [inline Dockerfile:45]: You can skip many of these when building a dedicated server.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+LABEL maintainer="Santhosh Janardhanan <santhoshj@gmail.com>" \
+    description="Dockerfile to build and run the latest OpenTTD from source."
+ARG APPVER=14.1
+# --------------------------------------------------------
+# 2) Environment Variables for Server Configuration
+# --------------------------------------------------------
+# These can be overridden at 'docker run' or in a docker-compose.yml
+ENV SERVER_NAME="Docker OpenTTD" \
+    SERVER_PASSWORD="" \
+    COMPANY_PASSWORD="" \
```
Comment (rubidium42) [inline Dockerfile:15]: If you're creating a server, how can you set a company password?

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    START_YEAR="1950" \
+    PORT="3979"
```
Comment (rubidium42) [inline Dockerfile:17]: None of these variables seem to be used anywhere else.

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,155 @@
+# --------------------------------------------------------
+# 1) Base Image
+# --------------------------------------------------------
+FROM ubuntu:22.04
+
+LABEL maintainer="Santhosh Janardhanan <santhoshj@gmail.com>" \
+    description="Dockerfile to build and run the latest OpenTTD from source."
+ARG APPVER=14.1
```
Comment (rubidium42) [inline Dockerfile:8]: You cannot hardcode a version and call it 'Latest'. As soon the hardcoded version won't be 'Latest' any more.

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+# Pull linux-syscall-support submodule
+RUN ./configure && make && \
+    make install
+# 'make install' is not provided by breakpad, so we do it manually:
+# mkdir -p /usr/local/include/breakpad && \
+# mkdir -p /usr/local/lib && \
+# cp -r src/* /usr/local/include/breakpad/ && \
+# find . -name '*.a' -exec cp {} /usr/local/lib/ \; && \
+# Cleanup
+# cd / && rm -rf /tmp/breakpad
```
Comment (rubidium42) [inline Dockerfile:77]: Why retain commented out code, especially when you state 'make install' is not provided and you are using 'make install' just before?

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+# --------------------------------------------------------
+# 7) Default Command (Dedicated Server)
+# --------------------------------------------------------
+# We use 'sh -c' so that environment variables are expanded at runtime.
+# -D = Dedicated (no GUI)
+# -x <year> = Start year
+# -n 0.0.0.0:<port> = Listen on all interfaces at <port>
+#   (We combine that with the server name using a config approach or '-N' if supported.)
+#
+# In modern OpenTTD builds, there's no direct param for server_name; it typically reads openttd.cfg
```
Comment (rubidium42) [inline Dockerfile:138]: Why 'In modern ...'? As far as I am aware there never was a parameter for server name. And since this is for inclusion into vanilla OpenTTD, anything that isn't in vanilla OpenTTD shouldn't be mentioned. Or those features should be added to vanilla OpenTTD.

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+
+
+# --------------------------------------------------------
+# 7) Default Command (Dedicated Server)
+# --------------------------------------------------------
+# We use 'sh -c' so that environment variables are expanded at runtime.
+# -D = Dedicated (no GUI)
+# -x <year> = Start year
+# -n 0.0.0.0:<port> = Listen on all interfaces at <port>
```
Comment (rubidium42) [inline Dockerfile:135]: `-n 0.0.0.0:<port>` does not listen on all interfaces; it definitely excludes IPv6 interfaces. With `-n [::]:<port>`you'll listen to both IPv4 and IPv6.

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+#   (We combine that with the server name using a config approach or '-N' if supported.)
+#
+# In modern OpenTTD builds, there's no direct param for server_name; it typically reads openttd.cfg
+# or is set by environment variables in official Docker images. However, you could:
+#   - rely on openttd.cfg, or
+#   - pass '-N "MyServer"' if your build supports it.
+#
+# For demonstration, we show a somewhat "hybrid" approach using '-n' for net config
+# and a sample '-x' for start year. Adjust to your needs or configure openttd.cfg.
```
Comment (rubidium42) [inline Dockerfile:144]: `-x` does not set the start year. 

Also the documentation seems to imply things are passed, but they are not. So you're actually describing some parameters that could be manually added to the command line. In that case, make it clear that they could be passed.

File: README.md [hunk 11]
```diff
@@ -3,18 +3,19 @@
...
-    - 2.1) [Multiplayer games](#21-multiplayer-games)
-    - 2.2) [Contributing to OpenTTD](#22-contributing-to-openttd)
-    - 2.3) [Reporting bugs](#23-reporting-bugs)
-    - 2.4) [Translating](#24-translating)
+  - 2.1) [Multiplayer games](#21-multiplayer-games)
+  - 2.2) [Contributing to OpenTTD](#22-contributing-to-openttd)
+  - 2.3) [Reporting bugs](#23-reporting-bugs)
+  - 2.4) [Translating](#24-translating)
 - 3.0) [Licensing](#30-licensing)
 - 4.0) [Credits](#40-credits)
```
Comment (rubidium42) [inline README.md:20]: Why does the readme need to be completely reformatted?
If there's a good reason for it, it should ideally be a separate commit/PR.

File: openttd.cfg-example [hunk 12]
```diff
@@ -0,0 +1,85 @@
+[server]
+# Password required for clients to join
+#server_password = "secretpassword"
+
+# Password for any company if used
+#company_password = "companypass"
```
Comment (rubidium42) [inline openttd.cfg-example:6]: A company password makes no sense in a dedicated server.

File: openttd.cfg-example [hunk 13]
```diff
@@ -0,0 +1,85 @@
+[server]
+# Password required for clients to join
+#server_password = "secretpassword"
+
+# Password for any company if used
+#company_password = "companypass"
+
+# Password for administration console (RCON)
+server_admin_password = "adminpass"
```
Comment (rubidium42) [inline openttd.cfg-example:9]: Server and admin passwords are not in openttd.cfg.

File: openttd.cfg-example [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+server_advertise = true
+
+# How many companies can exist in the game
+max_companies = 8
+
+# How many clients can be connected at once
+max_clients = 50
+
+# If true, the game starts paused, waiting for players
+start_paused = false
```
Comment (rubidium42) [inline openttd.cfg-example:24]: Do you mean 'pause_on_newgame'? Might not be what you want, as then you manually need to unpause.

File: openttd.cfg-example [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+# If false, the game continues running even if no players are connected
+pause_when_no_clients = false
```
Comment (rubidium42) [inline openttd.cfg-example:27]: This setting does not exist. Do you mean `min_active_clients = <n>`, where the game gets paused when there are less than `<n>` clients?

File: openttd.cfg-example [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+#company_password = "companypass"
+
+# Password for administration console (RCON)
+server_admin_password = "adminpass"
+
+# The port the server listens on for incoming connections (default is 3979)
+server_port = 3979
+
+# Whether or not to advertise this server (true/false)
+server_advertise = true
```
Comment (rubidium42) [inline openttd.cfg-example:15]: This setting does not exist any more. Do you mean 'server_game_type'?

File: openttd.cfg-example [hunk 17]
```diff
@@ -0,0 +1,85 @@
+[server]
```
Comment (rubidium42) [inline openttd.cfg-example:1]: There are no settings under `[server]` known to OpenTTD.

File: openttd.cfg-example [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+# If true, the game starts paused, waiting for players
+start_paused = false
+
+# If false, the game continues running even if no players are connected
+pause_when_no_clients = false
+
+
+[network]
+# The maximum time (in milliseconds) the server waits for a response before disconnecting
+network_max_tcp_clients = 20
```
Comment (rubidium42) [inline openttd.cfg-example:32]: As far as I am aware this setting does not exist. Even if it does, 20 ms is really short.

File: openttd.cfg-example [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+
+# The name that appears in the multiplayer server list
+server_name = "OpenTTD Server (v14.1) on Docker"
```
Comment (rubidium42) [inline openttd.cfg-example:35]: Does this need to be updated for each and every release?

File: openttd.cfg-example [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+# network_sync_enable = true
+
+
+[game_creation]
+starting_year = 1990
+map_x = 10
+map_y = 10
+
+[misc]
+# Language file (if available)
```
Comment (rubidium42) [inline openttd.cfg-example:48]: Game doesn't start when there's no language available...

File: openttd.cfg-example [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+[network]
+# The maximum time (in milliseconds) the server waits for a response before disconnecting
+network_max_tcp_clients = 20
+
+# The name that appears in the multiplayer server list
+server_name = "OpenTTD Server (v14.1) on Docker"
+
+# Extra lines you can tweak, e.g.:
+# network_frame_freq = 2
+# network_sync_enable = true
```
Comment (rubidium42) [inline openttd.cfg-example:39]: Either the settings aren't named this, or they don't exist at all as settings.

File: Dockerfile [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+# FROM ubuntu:22.04 AS builder
+
+# Install system dependencies and tools for building OpenGFX
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    git \
+    python3 \
+    python3-pip \
+    make \
+    imagemagick \
+    optipng grfcodec\
```
Comment (rubidium42) [inline Dockerfile:113]: Why no newline between packages here, when there is one for all other cases?

**Comment by TrueBrain** [comment]
For your context, https://github.com/OpenTTD/OpenTTD/issues/8663#issuecomment-776515496 .

Although we really appreciate your work, so far we always declined offers to integrate Dockerfiles in our upstream, as there are too many choices and variations possible. Picking one of them seems a bit weird, so thus far we left it to the community to arrange that. And they have supplied. There are multiple examples of how to run OpenTTD in Docker. See the above link for more detail.

@glx22 / @rubidium42 : given your in-depth comments on this PR, does this mean you are in favour of integrating one variant? As by the above link, personally I don't mind as much, as long as we do it knowingly :) I can also fully understand that the comments were purely a: I see things wrong in a PR, but I haven't thought about the PR in its context :D Which ofc is also perfectly fine :)

**Comment by rubidium42** [comment]
> @rubidium42 : given your in-depth comments on this PR, does this mean you are in favour of integrating one variant?

I wasn't aware (or couldn't remember) of any earlier PRs related to Docker. I'm indifferent to adding a variant. On one hand it makes it easier to deploy a dedicated server, but on the other hand it uses the same server name so people need to modify things manually in any case, which would favour having your own/separate git repository and then it would be better if it were not in the main repository. Maybe a separate repository would be better, though those already exist given the PR TrueBrain linked; I haven't looked at the quality of those docker configurations though.
But then... I haven't used docker yet, so I can't really say what is useful and what isn't regarding the use of git in there.
